File: Medium/3Sum/three_sum.py
from typing import List

def threeSum(nums: List[int]) -> List[List[int]]:
    nums.sort()
    result = []
    n = len(nums)

    for i in range(n - 2):
        if i > 0 and nums[i] == nums[i - 1]:
            continue  # Skip duplicate values for the first number

        left, right = i + 1, n - 1
        while left < right:
            total = nums[i] + nums[left] + nums[right]
            if total < 0:
                left += 1
            elif total > 0:
                right -= 1
            else:
                result.append([nums[i], nums[left], nums[right]])
                while left < right and nums[left] == nums[left + 1]:
                    left += 1  # Skip duplicate values for the second number
                while left < right and nums[right] == nums[right - 1]:
                    right -= 1  # Skip duplicate values for the third number
                left += 1
                right -= 1

    return result

if __name__ == "__main__":
    print("Testing threeSum function:")
    print("For the input nums = [-1,0,1,2,-1,-4], the output is:", threeSum([-1,0,1,2,-1,-4]))  # Output: [[-1,-1,2],[-1,0,1]]
    print("For the input nums = [0,1,1], the output is:", threeSum([0,1,1]))  # Output: []
    print("For the input nums = [0,0,0], the output is:", threeSum([0,0,0]))  # Output: [[0,0,0]]

File: Medium/3Sum/three_sum.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
    public:
        vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> result;
        sort(nums.begin(), nums.end()); //Sort the array first
        int n = nums.size();
        for(int i = 0; i < n - 2; i++){
            if(i > 0 && nums[i] == nums[i - 1]){
                continue; //Skip duplicates
            }
            int left = i + 1;
            int right = n - 1;
            while(left < right){
                int sum = nums[i] + nums[left] + nums[right];
                if(sum == 0){
                    result.push_back({nums[i], nums[left], nums[right]}); //Found a triplet
                    while(left < right && nums[left] == nums[left + 1]){
                        left++;
                    }
                    while(left < right && nums[right] == nums[right - 1]){
                        right--;
                    }
                    left++;
                    right--;
                } else if(sum < 0){
                    left++;
                } else{
                    right--;
                }
            }
        }
        return result;
    }
};

void printResult(const vector<vector<int>>& result){
    if (result.empty()) {
        cout << "[]" << endl;
        return;
    }
    for (const auto& triplet : result) {
        cout << "[";
        for (int i = 0; i < triplet.size(); i++) {
            cout << triplet[i];
            if (i < triplet.size() - 1) cout << ", ";
        }
        cout << "]" << endl;
    }
}

int main(){
    Solution solution1;
    vector<int> nums1 = {-1, 0, 1, 2, -1, -4};
    vector<vector<int>> result1 = solution1.threeSum(nums1);
    cout << "Triplets for nums1:" << endl;
    printResult(result1);
    vector<int> nums2 = {};
    vector<vector<int>> result2 = solution1.threeSum(nums2);
    cout << "Triplets for nums2:" << endl;
    printResult(result2);
    vector<int> nums3 = {0, 0, 0};
    vector<vector<int>> result3 = solution1.threeSum(nums3);
    cout << "Triplets for nums3:" << endl;
    printResult(result3);
    return 0;
}